ouroboros-network/src/Ouroboros/Network/Mux/Control.hs

```diff
@@ -1,61 +1,227 @@
 import           Data.Maybe (catMaybes)
+import           Data.Proxy (Proxy (..))
+import           Data.List (find)
+import           Data.Type.Equality
+import           GHC.TypeNats
+
+
+-- |
+-- Type family which maps version number to version fields.  Initial there are
+-- two instance, one in `Ouroboros.Network.NodeToNode` and
+-- `Ouroboros.Network.NodeToClient`.  First ten `0` - `9` are reserved for
+-- testing instances.
+--
+type family MuxVersion (n :: Nat)
+
+-- |
+-- Singletons used to pattern match version.
+--
+-- Adding a new version requires adding a constructor here, this will allow to
+-- pattern match on @'SomeVersion'@ and get access to it's second field of type
+-- @'MuxVersion' n@.
+--
+data SNat (n :: Nat) where
+  SNat0  :: SNat 0  -- ^ test mux version 0
+  SNat1  :: SNat 1  -- ^ test mux version 1
+
+  SNat10 :: SNat 10 -- ^ Node2Node mux protocol
+  SNat11 :: SNat 11 -- ^ Node2Client mux version
+
+asProxy :: SNat n -> Proxy n
+asProxy _ = Proxy
+
+versionNumber :: KnownNat n => SNat (n :: Nat) -> Word
+versionNumber = fromIntegral . natVal . asProxy
+
+-- |
+-- Existential wrapper for @'VersionField'@; it also carries singleton for @n@
+-- which allows to pattern match on constructors of @'MuxVersion' n@.
+--
+data SomeVersion where
+  SomeVersion
+    :: forall (n :: Nat).
+       ( KnownNat   n
+       , Serialise (MuxVersion n)
+       , Ord       (MuxVersion n)
+       , Eq        (MuxVersion n)
+       , Show      (MuxVersion n)
+       )
+    => SNat n       -- ^ singleton for (n :: Nat)
+    -> MuxVersion n -- ^ version fields
+    -> SomeVersion
+
+instance Eq SomeVersion where
+  (SomeVersion n v) == (SomeVersion n' v') =
+    case sameNat (asProxy n) (asProxy n') of
+      Just ref -> gcastWith ref (v == v')
+      Nothing  -> False
+
+instance Ord SomeVersion where
+  compare (SomeVersion n v) (SomeVersion n' v') =
+    case sameNat (asProxy n) (asProxy n') of
+      Just ref -> gcastWith ref (v `compare` v')
+      Nothing  -> natVal n `compare` natVal n'
+
+instance Show SomeVersion where
+  show (SomeVersion n v) = "SomeVersion " ++ show (natVal n) ++ " " ++ show v
+
+getDecoder :: SomeVersion -> SomeDecoder s
+getDecoder (SomeVersion n _) = SomeDecoder n decode
+
+-- |
+-- Existential wrapper for decoder; this is an internal type only used in this
+-- module.
+--
+data SomeDecoder s where
+  SomeDecoder
+    :: forall (n :: Nat) s.
+       ( KnownNat n
+       , Serialise (MuxVersion n)
+       , Ord (MuxVersion n)
+       , Eq (MuxVersion n)
+       , Show (MuxVersion n)
+       )
+    => SNat n
+    -> CBOR.Decoder s (MuxVersion n)
+    -> SomeDecoder s
+
+lookupDecoder :: Word -> [SomeVersion] -> Maybe (SomeDecoder s)
+lookupDecoder w vs = case find (\(SomeVersion n _) -> versionNumber n == w) vs of
+    Just sv -> Just (getDecoder sv)
+    Nothing -> Nothing
+
+encodeSomeVersion :: SomeVersion -> CBOR.Encoding
+encodeSomeVersion (SomeVersion n v) =
+       CBOR.encodeListLen 2
```

Comment:
In `encodeControlMsg` `MsgInitRsp` is encoded as a list of 3 fields, key, version number and version term. Encoding a version as a list with two elements doesn't match that.

Reply:
Well spotted, this means I should add a test for this :)